### PR TITLE
Unlimited undo/redo for Fl_Input_ and Fl_Text_Buffer (#558)

### PR DESCRIPTION
**Pull request by MatthiasWM**
- [x] for Fl_Input_
- [x] for Fl_Text_Buffer

### CODE REVIEW

**Comment by MatthiasWM** [comment]
I did some extensive testing with Fl_Text_Buffer and I think it behaves as users would expect it does. The concept of yank cuts, and when actions are merged, may be fine tuned, but I think this works 99.9% of the time.

In Fl_Input_, the code can be kept a lot simpler. Basic testing seems to do everything as expected.

File: src/Fl_Input.cxx [hunk 1]
```diff
@@ -278,7 +278,7 @@ int Fl_Input::kf_undo() {
 // Redo. (currently unimplemented.. toggles undo() instead)
 int Fl_Input::kf_redo() {
   if (readonly()) { fl_beep(); return 1; }
-  return kf_undo();                     // currently we don't support multilevel undo
+  return redo();                     // currently we don't support multilevel undo
 }
```
Comment (Albrecht-S) [inline src/Fl_Input.cxx:282]: The comment(s) above this "currently unimplemented" and in the line above "currently we don't support multilevel undo" appear to be outdated. Isn't this now implemented? I see that you changed kf_**undo** to **redo** but this looks like it was intended. Only the comments are questionable.

File: src/Fl_Input_.cxx [hunk 2]
```diff
@@ -186,7 +230,7 @@ double Fl_Input_::expandpos(
 /** \internal
   Marks a range of characters for update.
 
-  This call marks all characters from \p to the end of the
+  This call marks all characters from `p` to the end of the
```
Comment (Albrecht-S) [inline src/Fl_Input_.cxx:233]: Should be: `\p p` (or if you like: `\p 'p'`).

File: src/Fl_Input_.cxx [hunk 3]
```diff
@@ -947,13 +997,13 @@ int Fl_Input_::replace(int b, int e, const char* text, int ilen) {
...
-  Undoes previous changes to the text buffer.
+ Apply the current undo/redo operation
 
-  This call undoes a number of previous calls to replace().
+ It's up to undo() and redo() to push and po actions to and from the lists.
 
-  \return non-zero if any change was made.
-*/
-int Fl_Input_::undo() {
+ \return 1 if the current action cahnged any text.
```
Comment (Albrecht-S) [inline src/Fl_Input_.cxx:1004]: s/cahnged/changed/

File: src/Fl_Input_.cxx [hunk 4]
```diff
@@ -947,13 +997,13 @@ int Fl_Input_::replace(int b, int e, const char* text, int ilen) {
 }
 
 /**
-  Undoes previous changes to the text buffer.
+ Apply the current undo/redo operation
 
-  This call undoes a number of previous calls to replace().
+ It's up to undo() and redo() to push and po actions to and from the lists.
```
Comment (Albrecht-S) [inline src/Fl_Input_.cxx:1002]: s/po/pop/

File: src/Fl_Text_Buffer.cxx [hunk 5]
```diff
@@ -66,6 +66,23 @@ static int min(int i1, int i2)
 
 #endif
 
+/*
+ Undo/Redo is handled with Fl_Text_Undo_Action. The names of the class memebers
```
Comment (Albrecht-S) [inline src/Fl_Text_Buffer.cxx:70]: s/memebers/members/

File: src/Fl_Text_Buffer.cxx [hunk 6]
```diff
@@ -66,6 +66,23 @@ static int min(int i1, int i2)
...
+ relate to the original action.
+
+ Deleteing text will store the number of bytes deleted in `undocut`, and store
```
Comment (Albrecht-S) [inline src/Fl_Text_Buffer.cxx:73]: s/Deleteing/Deleting/

File: src/Fl_Text_Buffer.cxx [hunk 7]
```diff
@@ -66,6 +66,23 @@ static int min(int i1, int i2)
...
+ the deleted text in `undobuffer`. `undoat` is the insertion position.
+
+ Inserting text will store the number of bytes inserted in `undoinsert` and
+ `undoat` will point after the inserted text.
+
+ If text is deleted first and then text is inserted at the same position, it's
+ called a yankcut, and the number of bytes that were deleted is stored in
+ `undoyankcut`, again storing the deleted text in `undobuffer`.
+
+ If an undo action is run, text is deletd and inserted via the normal
```
Comment (Albrecht-S) [inline src/Fl_Text_Buffer.cxx:83]: s/deletd/deleted/

File: src/Fl_Text_Buffer.cxx [hunk 8]
```diff
@@ -102,6 +119,75 @@ class Fl_Text_Undo_Action {
...
+/*
+ Undo events are stored in a Last In - First Out stack.
+
+ Any insertion or deletion of text will either add to the current undo event
+ in mUndo, or generate a new undo event if cursor positions are not consecutive.
+ The previously current undo event will then be pushed to the undo list and
+ the redo event list is purged.
+
+ If the user calls undo(), the current undo event in mUndo will be run,
+ generatng a matching redo event in mUndo. The redo event is then pushed into
```
Comment (Albrecht-S) [inline src/Fl_Text_Buffer.cxx:137]: s/generatng/generating/

File: src/Fl_Text_Buffer.cxx [hunk 9]
```diff
@@ -102,6 +119,75 @@ class Fl_Text_Undo_Action {
...
+ the redo list, and the next undo event is popped form the undo list and made
```
Comment (Albrecht-S) [inline src/Fl_Text_Buffer.cxx:138]: s/form/from/

File: src/Fl_Text_Buffer.cxx [hunk 10]
```diff
@@ -102,6 +119,75 @@ class Fl_Text_Undo_Action {
...
+    } else
+      return NULL;
+  }
+
+  void clear() {
+    if (locked_) return;
+    if (list_) {
+      for (int i=0; i<list_size_; i++)
+        delete list_[i];
+      delete list_;
```
Comment (Albrecht-S) [inline src/Fl_Text_Buffer.cxx:182]: I suggest adding parentheses for clarity:
```
      for (int i=0; i<list_size_; i++) {
        delete list_[i];
      }
```

File: src/Fl_Text_Buffer.cxx [hunk 11]
```diff
@@ -459,51 +552,97 @@ void Fl_Text_Buffer::copy(Fl_Text_Buffer * fromBuf, int fromStart,
...
+  // save the current undo action and add an empty action to avoid generating yankcuts
+  Fl_Text_Undo_Action* action = mUndo;
+  mUndo = new Fl_Text_Undo_Action();
+
+  int ret = apply_undo(action, cursorPos);
+  delete action;
+
+  if (ret) {
+    // push the genearted undo action to the redo list
+    mRedoList->push(mUndo);
```
Comment (Albrecht-S) [inline src/Fl_Text_Buffer.cxx:616]: s/genearted/generated/
s/emty/empty/  (below this line)

File: src/Fl_Text_Buffer.cxx [hunk 12]
```diff
@@ -1218,10 +1357,20 @@ int Fl_Text_Buffer::insert_(int pos, const char *text)
 
   if (mCanUndo) {
     if (mUndo->undoat == pos && mUndo->undoinsert) {
+      // continue inserting text at the given cursor postion
```
Comment (Albrecht-S) [inline src/Fl_Text_Buffer.cxx:1360]: s/postion/position/

File: src/Fl_Text_Editor.cxx [hunk 13]
```diff
@@ -132,7 +132,9 @@ static struct {
   { FL_Page_Down, FL_CTRL|FL_SHIFT,         Fl_Text_Editor::kf_c_s_move   },
 //{ FL_Clear,     0,                        Fl_Text_Editor::delete_to_eol },
   { 'z',          FL_CTRL,                  Fl_Text_Editor::kf_undo       },
-  { '/',          FL_CTRL,                  Fl_Text_Editor::kf_undo       },
+  { 'z',          FL_CTRL|FL_SHIFT,         Fl_Text_Editor::kf_redo       }, // Windows users may be used to Ctr-Y
```
Comment (Albrecht-S) [inline src/Fl_Text_Editor.cxx:135]: s/Ctr-Y/Ctrl-Y/

Hmm, this comment is true, but what are the consequences? Do we ignore this fact?

**Review by Albrecht-S — COMMENTED** [review]
I'm sorry that I can only point at and correct some typos in docs and comments. I don't have the time to analyze the code and test, hence this may not be very helpful.

File: src/Fl_Text_Editor.cxx (hunk 13 above)
Comment (MatthiasWM) [reply src/Fl_Text_Editor.cxx]: I added this specifically for MS Windows (Screen Driver)

**Comment by MatthiasWM** [comment]
Thanks. Sorry for all the typos.

**Comment by MatthiasWM** [comment]
I decided to merge this. There may be minor issues, but I can't test every possible constellation. I tried some free-form editing and things seems well.

**Pull request merged**